File: src/scenesplit/video.py
"""Video loading and frame extraction module."""

from dataclasses import dataclass
from pathlib import Path
from typing import Iterator

import cv2
import numpy as np
from numpy.typing import NDArray

from scenesplit.constants import SUPPORTED_VIDEO_FORMATS, QualityPreset
from scenesplit.exceptions import (
    UnsupportedFormatError,
    VideoDecodeError,
    VideoNotFoundError,
)


@dataclass(frozen=True)
class VideoMetadata:
    """Metadata extracted from a video file."""

    path: Path
    width: int
    height: int
    fps: float
    frame_count: int
    duration_seconds: float
    codec: str


@dataclass(frozen=True)
class Frame:
    """A single video frame with metadata."""

    index: int
    timestamp_seconds: float
    data: NDArray[np.uint8]

    @property
    def timestamp_ms(self) -> int:
        """Timestamp in milliseconds."""
        return int(self.timestamp_seconds * 1000)


class VideoLoader:
    """Load and extract frames from video files.

    This class handles video file validation, opening, and frame extraction.
    All processing is done offline without any network access.
    """

    def __init__(self, path: str | Path) -> None:
        """Initialize the video loader.

        Args:
            path: Path to the video file.

        Raises:
            VideoNotFoundError: If the file does not exist.
            UnsupportedFormatError: If the file format is not supported.
            VideoDecodeError: If the video cannot be opened or decoded.
        """
        self.path = Path(path).resolve()
        self._validate_file()
        self._metadata: VideoMetadata | None = None
        self._cap: cv2.VideoCapture | None = None

    def _validate_file(self) -> None:
        """Validate the video file exists and has a supported format."""
        if not self.path.exists():
            raise VideoNotFoundError(str(self.path))

        if not self.path.is_file():
            raise VideoNotFoundError(f"Path is not a file: {self.path}")

        extension = self.path.suffix.lower().lstrip(".")
        if extension not in SUPPORTED_VIDEO_FORMATS:
            raise UnsupportedFormatError(str(self.path), extension)

    def _open(self) -> cv2.VideoCapture:
        """Open the video file and return the capture object."""
        cap = cv2.VideoCapture(str(self.path))
        if not cap.isOpened():
            raise VideoDecodeError(str(self.path), "Failed to open video file")
        return cap

    @property
    def metadata(self) -> VideoMetadata:
        """Get video metadata, loading it if necessary."""
        if self._metadata is None:
            cap = self._open()
            try:
                width = int(cap.get(cv2.CAP_PROP_FRAME_WIDTH))
                height = int(cap.get(cv2.CAP_PROP_FRAME_HEIGHT))
                fps = cap.get(cv2.CAP_PROP_FPS)
                frame_count = int(cap.get(cv2.CAP_PROP_FRAME_COUNT))
                fourcc = int(cap.get(cv2.CAP_PROP_FOURCC))

                # Decode fourcc to string
                codec = "".join([chr((fourcc >> 8 * i) & 0xFF) for i in range(4)])

                # Calculate duration
                duration = frame_count / fps if fps > 0 else 0.0

                self._metadata = VideoMetadata(
                    path=self.path,
                    width=width,
                    height=height,
                    fps=fps,
                    frame_count=frame_count,
                    duration_seconds=duration,
                    codec=codec.strip(),
                )
            finally:
                cap.release()

        return self._metadata

    def extract_frames(
        self,
        quality: QualityPreset = QualityPreset.BALANCED,
        progress_callback: callable | None = None,
    ) -> Iterator[Frame]:
        """Extract frames from the video at the specified sample rate.

        Args:
            quality: Quality preset determining sample rate.
            progress_callback: Optional callback(current, total) for progress.

        Yields:
            Frame objects containing frame data and metadata.

        Raises:
            VideoDecodeError: If frames cannot be read.
        """
        cap = self._open()
        try:
            sample_rate = quality.frame_sample_rate
            metadata = self.metadata
            total_frames = metadata.frame_count
            fps = metadata.fps

            frame_index = 0
            sampled_count = 0

            while True:
                ret, frame_data = cap.read()
                if not ret:
                    break

                # Only yield frames at the sample rate
                if frame_index % sample_rate == 0:
                    timestamp = frame_index / fps if fps > 0 else 0.0
                    yield Frame(
                        index=frame_index,
                        timestamp_seconds=timestamp,
                        data=frame_data,
                    )
                    sampled_count += 1

                frame_index += 1

                if progress_callback is not None:
                    progress_callback(frame_index, total_frames)

        finally:
            cap.release()

    def get_frame_at(self, index: int) -> Frame:
        """Get a specific frame by index.

        Args:
            index: Frame index (0-based).

        Returns:
            The Frame at the specified index.

        Raises:
            VideoDecodeError: If the frame cannot be read.
            IndexError: If the index is out of range.
        """
        metadata = self.metadata
        if index < 0 or index >= metadata.frame_count:
            raise IndexError(
                f"Frame index {index} out of range [0, {metadata.frame_count})"
            )

        cap = self._open()
        try:
            cap.set(cv2.CAP_PROP_POS_FRAMES, index)
            ret, frame_data = cap.read()
            if not ret:
                raise VideoDecodeError(
                    str(self.path), f"Failed to read frame at index {index}"
                )

            timestamp = index / metadata.fps if metadata.fps > 0 else 0.0
            return Frame(
                index=index,
                timestamp_seconds=timestamp,
                data=frame_data,
            )
        finally:
            cap.release()

    def __enter__(self) -> "VideoLoader":
        """Context manager entry."""
        return self

    def __exit__(self, exc_type, exc_val, exc_tb) -> None:
        """Context manager exit."""
        if self._cap is not None:
            self._cap.release()
            self._cap = None

File: src/video.rs
//! Video loading and frame extraction module.

use std::path::{Path, PathBuf};

use opencv::core::Mat;
use opencv::imgproc;
use opencv::prelude::*;
use opencv::videoio::{self, VideoCapture, VideoCaptureTraitConst};

use crate::config::QualityPreset;
use crate::error::{Error, Result, SUPPORTED_FORMATS};

/// Metadata extracted from a video file.
#[derive(Debug, Clone)]
#[allow(dead_code)]
pub struct VideoMetadata {
    pub path: PathBuf,
    pub width: u32,
    pub height: u32,
    pub fps: f64,
    pub frame_count: u32,
    pub duration_seconds: f64,
    pub codec: String,
}

/// A single video frame with metadata.
#[derive(Debug, Clone)]
pub struct Frame {
    pub index: usize,
    pub timestamp_seconds: f64,
    pub data: Vec<u8>,
    pub width: u32,
    pub height: u32,
}

impl Frame {
    /// Timestamp in milliseconds.
    #[allow(dead_code)]
    pub fn timestamp_ms(&self) -> u64 {
        (self.timestamp_seconds * 1000.0) as u64
    }
}

/// Video loader for extracting frames from video files.
pub struct VideoLoader {
    path: PathBuf,
    metadata: Option<VideoMetadata>,
}

impl VideoLoader {
    /// Create a new video loader.
    ///
    /// # Errors
    ///
    /// Returns an error if the file doesn't exist or has an unsupported format.
    pub fn new<P: AsRef<Path>>(path: P) -> Result<Self> {
        let path = path.as_ref().to_path_buf();
        Self::validate_file(&path)?;

        Ok(Self {
            path,
            metadata: None,
        })
    }

    fn validate_file(path: &Path) -> Result<()> {
        if !path.exists() {
            return Err(Error::VideoNotFound(path.to_path_buf()));
        }

        if !path.is_file() {
            return Err(Error::VideoNotFound(path.to_path_buf()));
        }

        let extension = path
            .extension()
            .and_then(|e| e.to_str())
            .map(|e| e.to_lowercase())
            .unwrap_or_default();

        if !SUPPORTED_FORMATS.contains(&extension.as_str()) {
            return Err(Error::UnsupportedFormat {
                path: path.to_path_buf(),
                extension,
            });
        }

        Ok(())
    }

    fn open_capture(&self) -> Result<VideoCapture> {
        let cap = VideoCapture::from_file(self.path.to_str().unwrap(), videoio::CAP_ANY)?;

        if !cap.is_opened()? {
            return Err(Error::VideoDecode {
                path: self.path.clone(),
                reason: "Failed to open video file".to_string(),
            });
        }

        Ok(cap)
    }

    /// Get video metadata, loading it if necessary.
    pub fn metadata(&mut self) -> Result<&VideoMetadata> {
        if self.metadata.is_none() {
            let cap = self.open_capture()?;

            let width = cap.get(videoio::CAP_PROP_FRAME_WIDTH)? as u32;
            let height = cap.get(videoio::CAP_PROP_FRAME_HEIGHT)? as u32;
            let fps = cap.get(videoio::CAP_PROP_FPS)?;
            let frame_count = cap.get(videoio::CAP_PROP_FRAME_COUNT)? as u32;
            let fourcc = cap.get(videoio::CAP_PROP_FOURCC)? as u32;

            // Decode fourcc to string
            let codec = (0..4)
                .map(|i| ((fourcc >> (8 * i)) & 0xFF) as u8 as char)
                .collect::<String>()
                .trim()
                .to_string();

            let duration = if fps > 0.0 {
                frame_count as f64 / fps
            } else {
                0.0
            };

            self.metadata = Some(VideoMetadata {
                path: self.path.clone(),
                width,
                height,
                fps,
                frame_count,
                duration_seconds: duration,
                codec,
            });
        }

        Ok(self.metadata.as_ref().unwrap())
    }

    /// Extract frames from the video at the specified sample rate.
    pub fn extract_frames<F>(
        &mut self,
        quality: QualityPreset,
        mut progress_callback: Option<F>,
    ) -> Result<Vec<Frame>>
    where
        F: FnMut(usize, usize),
    {
        let metadata = self.metadata()?.clone();
        let mut cap = self.open_capture()?;

        let sample_rate = quality.frame_sample_rate();
        let total_frames = metadata.frame_count as usize;
        let fps = metadata.fps;

        let mut frames = Vec::new();
        let mut frame_mat = Mat::default();
        let mut frame_index = 0usize;

        loop {
            let ret = cap.read(&mut frame_mat)?;
            if !ret || frame_mat.empty() {
                break;
            }

            // Only process frames at the sample rate
            if frame_index.is_multiple_of(sample_rate) {
                let timestamp = if fps > 0.0 {
                    frame_index as f64 / fps
                } else {
                    0.0
                };

                // Convert BGR to RGB
                let mut rgb_mat = Mat::default();
                imgproc::cvt_color_def(&frame_mat, &mut rgb_mat, imgproc::COLOR_BGR2RGB)?;

                // Get frame dimensions
                let width = rgb_mat.cols() as u32;
                let height = rgb_mat.rows() as u32;

                // Convert to Vec<u8>
                let data = mat_to_vec(&rgb_mat)?;

                frames.push(Frame {
                    index: frame_index,
                    timestamp_seconds: timestamp,
                    data,
                    width,
                    height,
                });
            }

            frame_index += 1;

            if let Some(ref mut cb) = progress_callback {
                cb(frame_index, total_frames);
            }
        }

        Ok(frames)
    }

    /// Get a specific frame by index.
    #[allow(dead_code)]
    pub fn get_frame_at(&mut self, index: usize) -> Result<Frame> {
        let metadata = self.metadata()?.clone();

        if index >= metadata.frame_count as usize {
            return Err(Error::VideoDecode {
                path: self.path.clone(),
                reason: format!(
                    "Frame index {} out of range [0, {})",
                    index, metadata.frame_count
                ),
            });
        }

        let mut cap = self.open_capture()?;
        cap.set(videoio::CAP_PROP_POS_FRAMES, index as f64)?;

        let mut frame_mat = Mat::default();
        let ret = cap.read(&mut frame_mat)?;

        if !ret || frame_mat.empty() {
            return Err(Error::VideoDecode {
                path: self.path.clone(),
                reason: format!("Failed to read frame at index {}", index),
            });
        }

        let timestamp = if metadata.fps > 0.0 {
            index as f64 / metadata.fps
        } else {
            0.0
        };

        // Convert BGR to RGB
        let mut rgb_mat = Mat::default();
        imgproc::cvt_color_def(&frame_mat, &mut rgb_mat, imgproc::COLOR_BGR2RGB)?;

        let width = rgb_mat.cols() as u32;
        let height = rgb_mat.rows() as u32;
        let data = mat_to_vec(&rgb_mat)?;

        Ok(Frame {
            index,
            timestamp_seconds: timestamp,
            data,
            width,
            height,
        })
    }

    /// Get the path to the video file.
    #[allow(dead_code)]
    pub fn path(&self) -> &Path {
        &self.path
    }
}

/// Convert an OpenCV Mat to a Vec<u8>.
fn mat_to_vec(mat: &Mat) -> Result<Vec<u8>> {
    let total = mat.total() * mat.channels() as usize;
    let mut data = vec![0u8; total];

    // Copy data from Mat
    let mat_data = mat.data_bytes()?;
    data.copy_from_slice(mat_data);

    Ok(data)
}
